tests/test.py

```diff
@@ -475,8 +502,131 @@ def recur(j1, j2, path, unordered=False):
             )
 
 
-class JSONTestCase(WithJSONTestCase, TestCase):
-    """TestCase subclass with JSON comparisons and test enable/disable flag
-    handling."""
+class TruTestCase(WithJSONTestCase, TestCase):
+    """TestCase subclass with several additions.
+
+    - JSON comparisons and golden-file handling.
+
+    - Dumps of remaining tasks on completion of each test. This can optionally
+      result in test failures if `TEST_TASKS_CLEANUP` is set.
+
+    - Dumps non-main running threads after test completion. This can be made to
+      result in test failure if `TEST_THREADS_CLEANUP` is set.
+
+    - Garbage collection subtest. This test optionally prints out the reference
+      path to the given object that was not garbage collected. This is enabled
+      with the `WITH_REF_PATH` environment variable.
+    """
+
+    def assertCollected(self, ref: weakref.ReferenceType[T], msg=None):
+        """Check that the object referenced by `ref` has been garbage
+        collected.
+
+        Optionally prints out the reference path to the object if it was not
+        garbage collected. This is enabled with the `WITH_REF_PATH` environment
+        variable.
+        """
+
+        gc.collect()
+
+        if msg is None:
+            msg = f"Object {ref} was not garbage collected."
+
+        obj = ref()
+
+        with self.subTest(part="garbage collection"):
+            self.assertTrue(obj is None, msg)
+
+        # Enable WITH_REF_PATH to see printout of why the given ref was not
+        # GC-ed.
+        if (
+            obj is not None
+            and os.environ.get(WITH_REF_PATH_VAR, None) is not None
+        ):
+            caller_globals = python_utils.caller_frame(offset=1).f_globals
+
+            with self.subTest(part="reference path"):
```

Comment:
what's the point of this `subTest`? Can't we just move this whole `if` block above L537 then have no subtests?

Reply:
Wanted the printout of the path to be under a seperate (sub)test than the GC collection (sub)test and only show up when the env var is set. There is also the possibility that there is something wrong with the path computation that may fail especially if the state of the interpreter is not settled at this point (i.e. other threads are still running and being activaly collected).